File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
  const map = new Map();

  // 1) 사람별 정보: parent(추천인), profit(누적 수익)
  for (let i = 0; i < enroll.length; i++) {
    map.set(enroll[i], {
      parent: referral[i] === '-' ? null : referral[i],
      profit: 0,
    });
  }

  // 2) 판매 수익 분배
  for (let i = 0; i < seller.length; i++) {
    let cur = seller[i];
    let money = amount[i] * 100; // 문제에서 1개당 100원

    while (cur !== null && money > 0) {
      const passUp = Math.floor(money * 0.1); // 10% (버림)
      const keep = money - passUp;            // 내 몫 90%

      map.get(cur).profit += keep;

      cur = map.get(cur).parent;
      money = passUp;
    }
  }

  // 3) enroll 순서대로 반환
  return enroll.map((name) => map.get(name).profit);
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def solution(enroll, referral, seller, amount):
    idx = {name: i for i, name in enumerate(enroll)}
    profits = [0] * len(enroll)

    for i, name in enumerate(seller):
        money = amount[i] * 100

        while name != "-" and money > 0:
            give = money // 10
            keep = money - give

            profits[idx[name]] += keep

            name = referral[idx[name]]
            money = give

    return profits